File: Problem/Luogu/P2184.cpp
//
// Created by kibi on 24-8-25.
//
#include <cctype>
#include <cstdio>
#include <system_error>
#include <sys/types.h>
#include <vector>
using namespace std;

template<typename T>
T read() {
	T s = 0, w = 1, c = getchar();
	while (!isdigit(c)) {
		if (c == '-') {
			w = -1;
		}
		c = getchar();
	}
	while (isdigit(c)) {
		s = s * 10 + c - '0';
		c = getchar();
	}
	return s * w;
}

class Opt{
public:
	uint q, l, r;
};

class Config {
public:
	uint n;
	vector<Opt> opts;
};

vector<uint> solve(const Config&);

int main() {
	auto n = read<uint>(), m = read<uint>();

	Config config{n, vector<Opt>(m)};
	for (auto &i : config.opts) {
		i = {read<uint>(), read<uint>(), read<uint>()};
	}

	auto ans = solve(config);
	for (auto &i : ans) {
		printf("%u\n", i);
	}

	return 0;
}

class BIT{
	vector<uint> value;
public:
	explicit BIT(const uint n) : value(n + 1) {}

	void add(uint pos, const uint val) {
		while(pos < value.size()) {
			value[pos] += val;
			pos += pos & -pos;
		}
	}

	uint query(uint pos) const {
		uint res = 0;
		while (pos) {
			res += value[pos];
			pos -= pos & -pos;
		}
		return res;
	}
};

vector<uint> solve(const Config &config) {
	vector<uint> res;
	BIT bit_left(config.n), bit_right(config.n);

	res.reserve(config.opts.size());

	for (auto &i : config.opts) {
		switch (i.q) {
			case 1:
				bit_left.add(i.l, 1);
			bit_right.add(i.r, 1);
			break;
			case 2:
				res.push_back(bit_left.query(i.r) - bit_right.query(i.l - 1));
			break;
			default:
				throw runtime_error("Error operation code!");
		}
	}

	return move(res);
}

File: Problem/Luogu/P2184.rs
use std::io::{Read, stdin};

fn main() {
    let config = read();

    let res = solve(config);

    for i in res {
        println!("{i}");
    }
}

struct Config {
    n: u32,
    opts: Vec<Opt>
}

enum Opt{
    M {l: u32, r: u32},
    Q {l: u32, r: u32},
}

fn getchar() -> char {
    let stdin = stdin();
    let mut handle = stdin.lock();

    let mut buffer = [0; 1];

    handle.read_exact(&mut buffer);

    let c = buffer[0] as char;

    if c.is_ascii() {
        return c;
    }
    std::process::exit(1);
    panic!("Error reading char!");
}

fn read_int() -> u32 {
    let mut c = getchar();
    let mut s = 0;
    while c < '0' || '9' < c {
        c = getchar();
    }
    while '0' <= c && c <= '9' {
        s = s * 10 + (c as u32 - '0' as u32);
        c = getchar();
    }
    s
}

fn read_str() -> String {
    let mut str = String::new();
    let mut c = getchar();
    while c < 'a' || 'z' < c {
        c = getchar();
    }
    while 'a' <= c && c <= 'z' {
        str.push(c);
        c = getchar();
    }
    str
}

fn read_char() -> char {
    let mut c = getchar();
    while c != 'Q' && c != 'R' && c != 'I' && (c < 'a' || c > 'z') {
        c = getchar();
    }
    c
}

fn read() -> Config {
    let mut input = String::new();
    stdin().read_line(&mut input).unwrap();
    let mut s = input.split_whitespace();
    let n = s.next().unwrap().parse().unwrap();
    let m = s.next().unwrap().parse().unwrap();
    input = String::new();

    let mut config = Config {n, opts:Vec::with_capacity(m)};
    for _ in 0..m {
        stdin().read_line(&mut input).unwrap();
        s = input.split_whitespace();
        let opt_type_input: u32 = s.next().unwrap().parse().unwrap();
        let opt;
        let l: u32 = s.next().unwrap().parse().unwrap();
        let r: u32 = s.next().unwrap().parse().unwrap();
        opt = match opt_type_input {
            1 => { Opt::M {l, r}},
            2 => { Opt::Q {l, r}},
            _ => panic!("Error operation types!"),
        };
        config.opts.push(opt);
        input = String::new();
    }

    config
}

struct BIT {
    value: Vec<u32>
}

impl BIT {
    fn new(n: usize) -> Self {
        let mut bit = Self{value: Vec::new()};
        bit.value.resize(n + 1, 0);
        bit
    }

    fn add (&mut self, mut pos: usize, val: u32) {
        while pos < self.value.len() {
            self.value[pos] += val;
            pos += pos & (pos ^ (pos - 1));
        }
    }

    fn query(&self, mut pos: usize) -> u32{
        let mut res:u32 = 0;
        while pos > 0 {
            res += self.value[pos];
            pos -= pos & (pos ^ (pos - 1));
        }
        res
    }
}

fn solve(config: Config) -> Vec<u32> {
    let mut res = Vec::new();
    let mut bit_left = BIT::new(config.n as usize);
    let mut bit_right = BIT::new(config.n as usize);

    for i in config.opts {
        match i {
            Opt::M {l, r} => {
                bit_left.add(l as usize, 1);
                bit_right.add(r as usize, 1);
            },
            Opt::Q {l, r} => {
                res.push(bit_left.query(r as usize) - bit_right.query(l as usize - 1));
            }
        }
    };

    res
}